gpbft/src/lib.rs

```diff
@@ -1,6 +1,194 @@
+    ///
+    /// * `instance` - The GPBFT instance number
+    /// * `round` - The current round number
+    /// * `step` - The current phase of the protocol
+    /// * `supplemental_data` - Additional data for this instance
+    /// * `value` - The agreed-upon ECChain
+    pub fn new(
+        instance: u64,
+        round: u64,
+        step: Phase,
+        supplemental_data: SupplementalData,
+        value: ECChain,
+    ) -> Self {
+        Payload {
+            instance,
+            round,
+            step,
+            supplemental_data,
+            value,
+        }
+    }
+}
+
+/// Represents a justification for a decision in the GPBFT consensus protocol
+pub struct Justification {
+    /// The payload that is signed by the signature
+    pub vote: Payload,
+    /// Indexes in the base power table of the signers (`bitset`)
+    pub signers: BitField,
+    /// BLS aggregate signature of signers
+    pub signature: Vec<u8>,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_payload_new() {
+        let instance = 1;
+        let round = 2;
+        let step = Phase::Commit;
+        let supplemental_data = SupplementalData {
+            commitments: keccak_hash::H256::zero(),
+            power_table: vec![],
+        };
+        let value = ECChain(Vec::new());
+
+        let payload = Payload::new(
+            instance,
+            round,
+            step,
+            supplemental_data.clone(),
+            value.clone(),
+        );
+
+        assert_eq!(payload.instance, instance);
+        assert_eq!(payload.round, round);
+        assert_eq!(payload.step, step);
+        assert_eq!(payload.supplemental_data, supplemental_data);
+        assert_eq!(payload.value, value);
+    }
+
+    #[test]
+    fn test_phase_ordering() {
+        assert!(Phase::Initial < Phase::Quality);
+        assert!(Phase::Quality < Phase::Converge);
+        assert!(Phase::Converge < Phase::Prepare);
+        assert!(Phase::Prepare < Phase::Commit);
+        assert!(Phase::Commit < Phase::Decide);
+        assert!(Phase::Decide < Phase::Terminated);
+    }
+
+    #[test]
+    fn test_phase_repr() {
+        assert_eq!(Phase::Initial as u8, 0);
+        assert_eq!(Phase::Quality as u8, 1);
+        assert_eq!(Phase::Converge as u8, 2);
+        assert_eq!(Phase::Prepare as u8, 3);
+        assert_eq!(Phase::Commit as u8, 4);
+        assert_eq!(Phase::Decide as u8, 5);
+        assert_eq!(Phase::Terminated as u8, 6);
+    }
+
+    #[test]
+    fn test_phase_display() {
+        assert_eq!(format!("{}", Phase::Initial), "INITIAL");
+        assert_eq!(format!("{}", Phase::Quality), "QUALITY");
+        assert_eq!(format!("{}", Phase::Converge), "CONVERGE");
+        assert_eq!(format!("{}", Phase::Prepare), "PREPARE");
+        assert_eq!(format!("{}", Phase::Commit), "COMMIT");
+        assert_eq!(format!("{}", Phase::Decide), "DECIDE");
+        assert_eq!(format!("{}", Phase::Terminated), "TERMINATED");
+    }
```

Comment:
Testing libraries' primitives doesn't bring any value. I'm unsure if it even increases coverage. We should assume the library works as intended.

Reply:
I'm actually unsure if the `UPPER` stuff will ever be critical, but keeping it compatible just in case.
Repr sequence is critical for sure.